// angular2-dynamic-component/src/DynamicComponent.ngfactory.ts
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
 /* tslint:disable */

import * as import0 from './DynamicComponent';
import * as import1 from '@angular/core/src/change_detection/change_detection';
import * as import2 from '@angular/core/src/linker/view_utils';
import * as import3 from '@angular/core/src/linker/view';
import * as import4 from '@angular/core/src/render/api';
import * as import5 from '@angular/core/src/linker/element';
import * as import6 from '@angular/core/src/di/injector';
import * as import7 from '@angular/core/src/linker/view_type';
import * as import8 from '@angular/core/src/linker/compiler';
import * as import9 from '@angular/http/src/http';
import * as import10 from '@angular/core/src/metadata/view';
import * as import11 from '@angular/core/src/linker/component_factory';
export class Wrapper_DynamicComponent {
  context:import0.DynamicComponent;
  changed:boolean;
  /*private*/ _componentType:any;
  /*private*/ _componentTemplate:any;
  /*private*/ _componentContext:any;
  /*private*/ _componentTemplateUrl:any;
  /*private*/ _componentDefaultTemplate:any;
  /*private*/ _componentRemoteTemplateFactory:any;
  /*private*/ _componentModules:any;
  constructor(p0:any,p1:any,p2:any,p3:any) {
    this.changed = false;
    this._componentType = import1.UNINITIALIZED;
    this._componentTemplate = import1.UNINITIALIZED;
    this._componentContext = import1.UNINITIALIZED;
    this._componentTemplateUrl = import1.UNINITIALIZED;
    this._componentDefaultTemplate = import1.UNINITIALIZED;
    this._componentRemoteTemplateFactory = import1.UNINITIALIZED;
    this._componentModules = import1.UNINITIALIZED;
    this.context = new import0.DynamicComponent(p0,p1,p2,p3);
  }
  check_componentType(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import2.checkBinding(throwOnChange,this._componentType,currValue))) {
      this.changed = true;
      this.context.componentType = currValue;
      this._componentType = currValue;
    }
  }
  check_componentTemplate(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import2.checkBinding(throwOnChange,this._componentTemplate,currValue))) {
      this.changed = true;
      this.context.componentTemplate = currValue;
      this._componentTemplate = currValue;
    }
  }
  check_componentContext(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import2.checkBinding(throwOnChange,this._componentContext,currValue))) {
      this.changed = true;
      this.context.componentContext = currValue;
      this._componentContext = currValue;
    }
  }
  check_componentTemplateUrl(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import2.checkBinding(throwOnChange,this._componentTemplateUrl,currValue))) {
      this.changed = true;
      this.context.componentTemplateUrl = currValue;
      this._componentTemplateUrl = currValue;
    }
  }
  check_componentDefaultTemplate(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import2.checkBinding(throwOnChange,this._componentDefaultTemplate,currValue))) {
      this.changed = true;
      this.context.componentDefaultTemplate = currValue;
      this._componentDefaultTemplate = currValue;
    }
  }
  check_componentRemoteTemplateFactory(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import2.checkBinding(throwOnChange,this._componentRemoteTemplateFactory,currValue))) {
      this.changed = true;
      this.context.componentRemoteTemplateFactory = currValue;
      this._componentRemoteTemplateFactory = currValue;
    }
  }
  check_componentModules(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import2.checkBinding(throwOnChange,this._componentModules,currValue))) {
      this.changed = true;
      this.context.componentModules = currValue;
      this._componentModules = currValue;
    }
  }
  detectChangesInternal(view:import3.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
var renderType_DynamicComponent_Host:import4.RenderComponentType = (null as any);
class _View_DynamicComponent_Host0 extends import3.AppView<any> {
  _el_0:any;
  /*private*/ _appEl_0:import5.AppElement;
  _DynamicComponent_0_5:Wrapper_DynamicComponent;
  constructor(viewUtils:import2.ViewUtils,parentInjector:import6.Injector,declarationEl:import5.AppElement) {
    super(_View_DynamicComponent_Host0,renderType_DynamicComponent_Host,import7.ViewType.HOST,viewUtils,parentInjector,declarationEl,import1.ChangeDetectorStatus.CheckAlways);
  }
  createInternal(rootSelector:string):import5.AppElement {
    this._el_0 = this.selectOrCreateHostElement('DynamicComponent',rootSelector,(null as any));
    this._appEl_0 = new import5.AppElement(0,(null as any),this,this._el_0);
    var compView_0:any = viewFactory_DynamicComponent0(this.viewUtils,this.injector(0),this._appEl_0);
    this._DynamicComponent_0_5 = new Wrapper_DynamicComponent(this.parentInjector.get('DynamicExtraModules'),this._appEl_0.vcRef,this.parentInjector.get(import8.Compiler),this.parentInjector.get(import9.Http));
    this._appEl_0.initComponent(this._DynamicComponent_0_5.context,([] as any[]),compView_0);
    compView_0.create(this._DynamicComponent_0_5.context,this.projectableNodes,(null as any));
    this.init(([] as any[]).concat([this._appEl_0]),[this._el_0],([] as any[]),([] as any[]));
    return this._appEl_0;
  }
  injectorGetInternal(token:any,requestNodeIndex:number,notFoundResult:any):any {
    if (((token === import0.DynamicComponent) && (0 === requestNodeIndex))) { return this._DynamicComponent_0_5.context; }
    return notFoundResult;
  }
  detectChangesInternal(throwOnChange:boolean):void {
    this._DynamicComponent_0_5.detectChangesInternal(this,this._el_0,throwOnChange);
    this.detectContentChildrenChanges(throwOnChange);
    this.detectViewChildrenChanges(throwOnChange);
  }
}
function viewFactory_DynamicComponent_Host0(viewUtils:import2.ViewUtils,parentInjector:import6.Injector,declarationEl:import5.AppElement):import3.AppView<any> {
  if ((renderType_DynamicComponent_Host === (null as any))) { (renderType_DynamicComponent_Host = viewUtils.createRenderComponentType('',0,import10.ViewEncapsulation.None,([] as any[]),{})); }
  return new _View_DynamicComponent_Host0(viewUtils,parentInjector,declarationEl);
}
export const DynamicComponentNgFactory:import11.ComponentFactory<import0.DynamicComponent> = new import11.ComponentFactory<import0.DynamicComponent>('DynamicComponent',viewFactory_DynamicComponent_Host0,import0.DynamicComponent);
const styles_DynamicComponent:any[] = ([] as any[]);
var renderType_DynamicComponent:import4.RenderComponentType = (null as any);
class _View_DynamicComponent0 extends import3.AppView<import0.DynamicComponent> {
  constructor(viewUtils:import2.ViewUtils,parentInjector:import6.Injector,declarationEl:import5.AppElement) {
    super(_View_DynamicComponent0,renderType_DynamicComponent,import7.ViewType.COMPONENT,viewUtils,parentInjector,declarationEl,import1.ChangeDetectorStatus.CheckAlways);
  }
  createInternal(rootSelector:string):import5.AppElement {
    const parentRenderNode:any = this.renderer.createViewRoot(this.declarationAppElement.nativeElement);
    this.init(([] as any[]),([] as any[]),([] as any[]),([] as any[]));
    return (null as any);
  }
}
export function viewFactory_DynamicComponent0(viewUtils:import2.ViewUtils,parentInjector:import6.Injector,declarationEl:import5.AppElement):import3.AppView<import0.DynamicComponent> {
  if ((renderType_DynamicComponent === (null as any))) { (renderType_DynamicComponent = viewUtils.createRenderComponentType('',0,import10.ViewEncapsulation.None,styles_DynamicComponent,{})); }
  return new _View_DynamicComponent0(viewUtils,parentInjector,declarationEl);
}
