// angular2-dynamic-component/DynamicComponentModule.js
"use strict";
var core_1 = require('@angular/core');
var DynamicComponent_1 = require('./DynamicComponent');
var DynamicDirective_1 = require("./DynamicDirective");
var DynamicBase_1 = require("./DynamicBase");
var DynamicComponentModule = (function () {
    function DynamicComponentModule() {
    }
    DynamicComponentModule.decorators = [
        { type: core_1.NgModule, args: [{
                    declarations: [
                        DynamicComponent_1.DynamicComponent,
                        DynamicDirective_1.DynamicDirective
                    ],
                    exports: [
                        DynamicComponent_1.DynamicComponent,
                        DynamicDirective_1.DynamicDirective
                    ],
                    providers: [
                        { provide: DynamicBase_1.DYNAMIC_TYPES.DynamicExtraModules, useValue: [] }
                    ]
                },] },
    ];
    /** @nocollapse */
    DynamicComponentModule.ctorParameters = [];
    return DynamicComponentModule;
}());
exports.DynamicComponentModule = DynamicComponentModule;
var DynamicComponentModuleFactory = (function () {
    function DynamicComponentModuleFactory() {
    }
    DynamicComponentModuleFactory.buildModule = function (dynamicExtraModules) {
        var DynamicComponentFactoryModule = (function () {
            function DynamicComponentFactoryModule() {
            }
            DynamicComponentFactoryModule.decorators = [
                { type: core_1.NgModule, args: [{
                            declarations: [
                                DynamicComponent_1.DynamicComponent,
                                DynamicDirective_1.DynamicDirective
                            ],
                            exports: [
                                DynamicComponent_1.DynamicComponent,
                                DynamicDirective_1.DynamicDirective
                            ],
                            providers: [
                                { provide: DynamicBase_1.DYNAMIC_TYPES.DynamicExtraModules, useValue: dynamicExtraModules }
                            ]
                        },] },
            ];
            /** @nocollapse */
            DynamicComponentFactoryModule.ctorParameters = [];
            return DynamicComponentFactoryModule;
        }());
        return DynamicComponentFactoryModule;
    };
    return DynamicComponentModuleFactory;
}());
exports.DynamicComponentModuleFactory = DynamicComponentModuleFactory;
